Fix "dir" in GJKDistance function

hongkai-dai opened the pull request:
The variable `dir` is overloaded in `GJKDistance` function, that it means both the point within the simplex that is closest to the origin, and the directional vector along which to compute the support mapping. One issue caused by this overload, is that when calling `extractClosestPoint`, we should pass in the closest point, but instead passed in the directional vector. In this PR I introduce a separate variable `closest_p` to represent the closest point.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/flexible-collision-library/fcl/290)
<!-- Reviewable:end -->

sherm1 commented:
Is this ready for review, @hongkai-dai ?

hongkai-dai commented:
Yes it is ready for review.

The error on Mac CI is likely due to libccd being compiled with single precision

SeanCurtis-TRI commented:
@sherm1 Want to split feature/platform review?

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-:-LDCsR1h14-xgOnuoAxa:bducp71)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+@seancurtis-tri  for feature review

I really like the directory structure in the `test` directory for unit tests. Thanks.

First pass done.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1330 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDCyCQc0GrQIy-nDkNk:-LDCyCQc0GrQIy-nDkNl:b-919gjq) ([raw file](https://github.com/flexible-collision-library/fcl/blob/1f508738ee8e20ff6193e98b5b77b2644b7dc269/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1330)):*
> ```objc
> 
> 
> /// change the libccd distance to add two closest points
> ```

BTW META: Based on the work you did, could you provide a function-level documentation for this function (and the others you touched to a sufficient degree to feel confident in your ability to document it)? You had to gain that understanding, it would be nice if future generations could benefit from that.

---

*[test/test_fcl_sphere_sphere.cpp, line 51 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDCysyk1ytMzFRUDMkZ:-LDCysyk1ytMzFRUDMk_:b-626eov) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L51)):*
> ```cpp
> // For two spheres 1, 2, sphere 1 has radius1, and is centered at point A, with
> // coordinate p_FA in some frame F; sphere 2 has radius2, and is centered at
> // point B, with coordinate p_FB in the same frame F. Compute the (signed)
> ```

BTW "Compute the (optionally signed) distance..."

---

*[test/test_fcl_sphere_sphere.cpp, line 85 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD-l3iDYyjJhiMF45e:-LDD-l3iDYyjJhiMF45f:b-3cmil8) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L85)):*
> ```cpp
>     S radius1, S radius2, const fcl::Vector3<S>& p_FA,
>     const fcl::Vector3<S>& p_FB, fcl::GJKSolverType solver_type,
>     bool enable_nearest_points, bool enable_signed_distance,
> ```

BTW Neither of these booleans appear to be anything *other* than true in this test. (Applies across this entire file).

---

*[test/test_fcl_sphere_sphere.cpp, line 103 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD1_3YBlIiwQrN_6Pu:-LDD1_3YBlIiwQrN_6Pv:b-nqppr0) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L103)):*
> ```cpp
>   const S radius2 = 0.6;
>   const fcl::Vector3<S> p_FA(0, 0, 0);
>   const fcl::Vector3<S> p_FB(1.2, 0, 0);
> ```

BTW It might be worth creating arrays of relative positions. Right now, one at the origin and one displaced along a single axis may be less exhausting than we'd really want.

---

*[test/test_fcl_sphere_sphere.cpp, line 105 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD0Cu-8dFS1VzoMN-w:-LDD0Cu-8dFS1VzoMN-x:bnbywb2) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L105)):*
> ```cpp
>   const fcl::Vector3<S> p_FB(1.2, 0, 0);
> 
>   const fcl::Vector3<S> p1_expected(0.5, 0, 0);
> ```

FYI If you expressed these in terms of your parameters it would be less brittle.

Then you could position the spheres in non-axis-aligned positions and still get the right answers.

---

*[test/test_fcl_sphere_sphere.cpp, line 106 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD0os34WqJsieP3o-Y:-LDD0os34WqJsieP3o-Z:bqebr2r) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L106)):*
> ```cpp
> 
>   const fcl::Vector3<S> p1_expected(0.5, 0, 0);
>   const fcl::Vector3<S> p2_expected(-0.6, 0, 0);
> ```

BTW Also, given notation `p_FA`, you should also do `p_ANa_expected` and `p_BNb_expected` where `Na` and `Nb` are the nearest points on spheres A and B, respectively. That makes them the position vectors from the sphere origin to its nearest point, measured and expressed in the sphere's frame.

Ultimately, what I *really* want is a clear statement that the answers are in the sphere's local frames.

---

*[test/test_fcl_sphere_sphere.cpp, line 107 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD1QiQB8rjJkak_Ldw:-LDD1QiQB8rjJkak_Ldx:bqswsos) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L107)):*
> ```cpp
>   const fcl::Vector3<S> p1_expected(0.5, 0, 0);
>   const fcl::Vector3<S> p2_expected(-0.6, 0, 0);
>   for (const fcl::GJKSolverType solver_type :
> ```

BTW A lot of the other tests in fcl separate the solvers into different unit tests. Without good reason, I'd propose we do the same.

---

*[test/test_fcl_sphere_sphere.cpp, line 121 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD6XBND9IsxI28KR7V:-LDD6XBND9IsxI28KR7W:b-i53luq) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L121)):*
> ```cpp
> }
> 
> GTEST_TEST(FCL_SPHERE_SPHERE, Separating_Spheres) {
> ```

BTW It should be more clear that the tests you've created are only testing *signed* distance. 

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 67 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD3PeL-vjOGzoIcyx4:-LDD3PeL-vjOGzoIcyx5:ba5a1vp) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L67)):*
> ```cpp
>     tf1.translation() = center1;
>     tf2.translation() = center2;
>     void* o1 = GJKInitializer<S, fcl::Sphere<S>>::createGJKObject(s1, tf1);
> ```

BTW These two lines constitute a memory leak; there should be an accompanying invocation of: `GJKInitializer<S, fcl::Sphere<S>>::deleteGJKObject(o1);`.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 89 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD43rd0tpnmQQIYuM5:-LDD43reEdpYQ82ikZLv:b-ed6e8s) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L89)):*
BTW You've got two typos and some slightly awkward grammar. So, I simply propose reworking the error:

> This test cannot be used for spheres whose centers are coincident.

You might also consider doing a:

```
GTEST_FAIL() << "Sphere centers are functionally coincident:"
             << \n\tp_FA: " << p_FA.transpose()
             << \n\tp_FB: " << p_FB.transpose();
```

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 100 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD506u-7JF_tTN29Hb:-LDD506u-7JF_tTN29Hc:bn52fk2) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L100)):*
> ```cpp
>   };
> 
>   CheckSignedDistance(radius1, radius2, center1, center2, tol);
> ```

BTW The value of swapping the order of the spheres in this test isn't worth as much as in `test_sphere_sphere.cc`. The reason is that the test doesn't confirm that these both produce the *same* answer, only that the internals of `CheckSignedDistance` are consistent.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 107 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD5NF34StXsDP7d_sp:-LDD5NF34StXsDP7d_sq:buurnne) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L107)):*
> ```cpp
> template <typename S>
> void TestNonCollidingSphereGJKSignedDistance(S tol) {
>   TestSphereToSphereGJKSignedDistance<S>(0.5, 0.5, Vector3<S>(0, 0, 0),
> ```

BTW As I commented on the other tests, I'd like to see an array of scenarios  rather than just this one.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 114 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD5Zxx598uvVElnW_f:-LDD5Zxx598uvVElnW_g:b-236n08) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L114)):*
> ```cpp
> 
> GTEST_TEST(FCL_GJKSignedDistance, sphere_sphere) {
>   TestNonCollidingSphereGJKSignedDistance<double>(1E-14);
> ```

BTW you can use `constants<S>::eps_78()` or `constants<S>::eps()` (probably the former) instead of these hard-coded values. Then it will equally apply if `S` is ever `AutoDiffXd`.

In fact, you could hard-code that tolerance *inside* the test rather than passing it in here.

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-:-LDCwdWP7IRIKSLkrJRF:b-84nd7j)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1330 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDCyCQc0GrQIy-nDkNk:-LDDN3LkBBNU9kVsaJuv:b-yh16l6) ([raw file](https://github.com/flexible-collision-library/fcl/blob/1f508738ee8e20ff6193e98b5b77b2644b7dc269/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1330)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW META: Based on the work you did, could you provide a function-level documentation for this function (and the others you touched to a sufficient degree to feel confident in your ability to document it)? You had to gain that understanding, it would be nice if future generations could benefit from that.
</blockquote></details>

Done. Good call.

---

*[test/test_fcl_sphere_sphere.cpp, line 51 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDCysyk1ytMzFRUDMkZ:-LDDPEOg5YdaF-wN_S3A:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L51)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW "Compute the (optionally signed) distance..."
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 85 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD-l3iDYyjJhiMF45e:-LDDPGex79tlV0qE1_HT:b-y6ds4f) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L85)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Neither of these booleans appear to be anything *other* than true in this test. (Applies across this entire file).
</blockquote></details>

That is true. By setting both these booleans to true, I test the function `_ccdDist` in `gjk_libccd-inl.h`, which contains the bug being fixed in this PR. I add a TODO here, to test with these booleans to false. I will add the test cases in the future PR.

---

*[test/test_fcl_sphere_sphere.cpp, line 103 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD1_3YBlIiwQrN_6Pu:-LDDPvgRAoWgDcPsr1OW:b-jpjtc5) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L103)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It might be worth creating arrays of relative positions. Right now, one at the origin and one displaced along a single axis may be less exhausting than we'd really want.
</blockquote></details>

Done. Good call.

I added more tests. But the precision for the new tests are not as good as 1E-14. They are now about 1E-3.

---

*[test/test_fcl_sphere_sphere.cpp, line 106 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD0os34WqJsieP3o-Y:-LDDnjpz8DSaps4hJ5-r:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L106)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Also, given notation `p_FA`, you should also do `p_ANa_expected` and `p_BNb_expected` where `Na` and `Nb` are the nearest points on spheres A and B, respectively. That makes them the position vectors from the sphere origin to its nearest point, measured and expressed in the sphere's frame.

Ultimately, what I *really* want is a clear statement that the answers are in the sphere's local frames.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 107 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD1QiQB8rjJkak_Ldw:-LDDxbiNAWk1yoqhQZ0e:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L107)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW A lot of the other tests in fcl separate the solvers into different unit tests. Without good reason, I'd propose we do the same.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 121 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD6XBND9IsxI28KR7V:-LDDxijGA3uM8kyVcKNf:b32zaal) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L121)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW It should be more clear that the tests you've created are only testing *signed* distance. 
</blockquote></details>

Done, I added the test with `enable_signed_distance = false`.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 67 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD3PeL-vjOGzoIcyx4:-LDDzwgH03_r-80Ad6E1:b-376vyo) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L67)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW These two lines constitute a memory leak; there should be an accompanying invocation of: `GJKInitializer<S, fcl::Sphere<S>>::deleteGJKObject(o1);`.
</blockquote></details>

Done. Good catch!

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 89 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD43rd0tpnmQQIYuM5:-LDE-S67-kSPXvnMuCzQ:bnputky) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L89)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW You've got two typos and some slightly awkward grammar. So, I simply propose reworking the error:

> This test cannot be used for spheres whose centers are coincident.

You might also consider doing a:

```
GTEST_FAIL() << "Sphere centers are functionally coincident:"
             << \n\tp_FA: " << p_FA.transpose()
             << \n\tp_FB: " << p_FB.transpose();
```
</blockquote></details>

Done. Thanks

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 100 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD506u-7JF_tTN29Hb:-LDE-lD6EZgxdbP9PNFg:b372ijy) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L100)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The value of swapping the order of the spheres in this test isn't worth as much as in `test_sphere_sphere.cc`. The reason is that the test doesn't confirm that these both produce the *same* answer, only that the internals of `CheckSignedDistance` are consistent.
</blockquote></details>

Without the bug fix on `_ccdDist`, one of these checks would fail. So I think it still makes sense to test both orders here?

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 107 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD5NF34StXsDP7d_sp:-LDE2taA4KY8yup6xSYl:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L107)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW As I commented on the other tests, I'd like to see an array of scenarios  rather than just this one.
</blockquote></details>

Done.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 114 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD5Zxx598uvVElnW_f:-LDE2z6K8o6XCwOH0En7:b1njpre) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L114)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW you can use `constants<S>::eps_78()` or `constants<S>::eps()` (probably the former) instead of these hard-coded values. Then it will equally apply if `S` is ever `AutoDiffXd`.

In fact, you could hard-code that tolerance *inside* the test rather than passing it in here.
</blockquote></details>

I pass it in here as I thought the tolerance would depend on the type of `S`. Actually with more tests, the tolerance drops to 1E-3.

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> with a few BTWs, ptal.

---

Reviewed 4 of 7 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1160 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEK9Vw2zUMmP6lTio_:-LDEK9Vx3-EFvJRcWl4N:b26rulg) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1160)):*
> ```objc
> // the distance. For more info about GJK algorithm, the readers can refer to 
> // section 2.3 of Real-time Collision Detection with Implicit Objects by Leif
> // Olvang, 2010
> ```

FYI you could use backticks for variables rather than `@p` -- somewhat easier to read and would also work in Doxygen. (Irrelevant for this non-doxygen comment)

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1171 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDELcj-8lLlLxru7VT3:-LDELcj-8lLlLxru7VT4:bhbonzk) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1171)):*
> ```objc
>                         // origin.
>   ccd_vec3_t dir; // direction vector
>   ccd_real_t dist, last_dist = CCD_REAL_MAX;
> ```

BTW consider moving these declarations down to just prior to their use and in the loop scope where possible. I had trouble figuring out what was used where by scanning the code. I believe `iterations` is just a local loop variable so can be moved to the for statement. Looks like only `closest_p` needs to be defined at the top?

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1350 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEMxpY0ih0blF-OY92:-LDEMxpY0ih0blF-OY93:b-ezm3oq) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1350)):*
> ```objc
>     return -CCD_ONE;
> 
>   return _ccdDist(obj1, obj2, ccd, &simplex, p1, p2);
> ```

BTW so great to see all this duplicate code deleted! Thanks!

---

*[test/test_fcl_sphere_sphere.cpp, line 35 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDENDqHDkC6alJkY4aC:-LDENDqHDkC6alJkY4aD:bjybg5x) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L35)):*
> ```cpp
>  */
> 
> /** @author Hongkai Dai <hongkai.dai@tri.global>*/
> ```

FYI could add a space before closing comment.

---

*[test/test_fcl_sphere_sphere.cpp, line 60 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDENTanAwLfe362WJEo:-LDENTanAwLfe362WJEp:bofoj2m) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L60)):*
> ```cpp
>                               bool enable_signed_distance,
>                               fcl::DistanceResult<S>* result) {
>   fcl::Transform3<S> X_FA, X_FB;
> ```

BTW consider a brief explanation of this notation and a link to Drake's documentation for it: http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html

Likely FCL programmers are not familiar with it yet.

---

*[test/test_fcl_sphere_sphere.cpp, line 180 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEOkIVE9veSRoCuxRv:-LDEOkIVE9veSRoCuxRw:b-pxf3jy) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L180)):*
> ```cpp
> 
> GTEST_TEST(FCL_SPHERE_SPHERE, Separating_Spheres_LIBCCD) {
>   TestSeparatingSpheres<double>(1E-3, fcl::GJKSolverType::GST_LIBCCD);
> ```

BTW Yikes! Is that really the best we can hope for with libccd?

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-:-LDEQsl204-Cz71q3quP:b-682olt)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Second pass

---

Reviewed 1 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1157 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHOkq9DqlFcPJA0IuZ:-LDHOkq9DqlFcPJA0Iu_:b-p8cuvt) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1157)):*
BTW I have a bunch of concerns about this documentation

1. p1 and p2 are expressed in which frame? (Comment applies below as well).
2. The return value is undocumented
3. The relationship between the objects and the simplex isn't explained.
4. The semantics of the simplex isn't clear -- is it properly initialized upon invocation? Or is it an output value?

I know it's asking a lot, but if this were elevated to a Drake-quality piece of documentation, we'd all benefit.

Something along the lines of:

> Computes the distance between two non-penetrating convex objects, `obj1` and `obj2` based on the warm-started witness `simplex`, returning the distance and nearest points on each object.
@param obj1     The first convex object.
@param obj2     The second convex object.
@param ccd      The libccd configuration.
@param simplex   A witness to the objects' separation generated by the GJK algorithm. NOTE: the simplex is not necessarily sufficiently refined to report the *actual* distance and may be further refined in this method.
@param p1       If the objects are non-penetrating, the point on the surface of `obj1` closest to `obj2` (expressed in the world frame).
@param p2       If the objects are non-penetrating, the point on the surface of `obj2` closest to `obj1` (expressed in the world frame).
@returns The minimum distance between the two objects. If they are penetrating, -1 is returned.

(With appropriate `//` injected at appropriate line width, of course.)

Unknowns: What happens if the simplex is not initialized? Or it's lying?

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1339 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHP_pY4uPAbZtKqK_B:-LDHP_pY4uPAbZtKqK_C:b-t1mls8) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1339)):*
> ```objc
> 
> 
> // Computes the (unsigned) distance between two geometric objects @p obj1 and @p
> ```

BTW I'd modify this to reflect the comment on the other function. The difference between the two is that this does the warm starting of the libccd GJK implementation.

---

*[test/test_fcl_sphere_sphere.cpp, line 49 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHjNxQBqydRx57skZ8:-LDHjNxQBqydRx57skZ9:bufypb5) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L49)):*
> ```cpp
> #include "fcl/narrowphase/distance_result.h"
> 
> // For two spheres 1, 2, sphere 1 has radius1, and is centered at point A, with
> ```

FYI I just realized: if you had spheres A and B and radius_A and radius_B, that makes p_FB and p_FA fit better.

---

*[test/test_fcl_sphere_sphere.cpp, line 116 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmRrY-eJALYDQKY-W:-LDHmRrY-eJALYDQKY-X:bynmg5s) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L116)):*
> ```cpp
>     min_distance = (sphere1.p_FC - sphere2.p_FC).norm() - sphere1.radius -
>                    sphere2.radius;
>     const fcl::Vector3<S> AB = (sphere1.p_FC- sphere2.p_FC).normalized();
> ```

BTW missing space before `-` operator.

---

*[test/test_fcl_sphere_sphere.cpp, line 117 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmWN67BGJK5AFIJgm:-LDHmWN67BGJK5AFIJgn:b-xnj2ew) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L117)):*
> ```cpp
>                    sphere2.radius;
>     const fcl::Vector3<S> AB = (sphere1.p_FC- sphere2.p_FC).normalized();
>     p_S1P1 = -AB * sphere1.radius;
> ```

FYI *slightly* more efficient to negate the radius. But yours reads better. :)

---

*[test/test_fcl_sphere_sphere.cpp, line 123 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmaTb0L_IXDTLuG1A:-LDHmaTb0L_IXDTLuG1B:bti55jq) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L123)):*
> ```cpp
>   SphereSpecification<S> sphere2;
>   S min_distance;
>   // The closest points are expressed in the sphere body frame.
> ```

BTW These three comments are slightly redundant. The first one includes the next two. So, I'd suggest one of two options:

1. Delete the last two comments, or
2. delete the first comment and then move the last comment to between the two Vector declarations (i.e., comment, object, comment, object.)

---

*[test/test_fcl_sphere_sphere.cpp, line 143 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHnHi77nFbCXuFLsT-:-LDHnHi77nFbCXuFLsT0:bcolnem) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L143)):*
> ```cpp
>     for (int j = i + 1; j < static_cast<int>(spheres.size()); ++j) {
>       SphereSphereDistance<S> sphere_sphere_distance(spheres[i], spheres[j]);
>       if (sphere_sphere_distance.min_distance > 0) {
> ```

BTW This is a dangerous test -- this should be a `ASSERT_GT(min_distance, 0)`. Otherwise, the spheres could be changed to be in collision and *no* testing would take place. Obviously, this will require you to guarantee that none of the sphere pairs are colliding.  :)

---

*[test/test_fcl_sphere_sphere.cpp, line 180 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEOkIVE9veSRoCuxRv:-LDHo1dD8jhNffSHE0pX:b-ngkrj5) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L180)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Yikes! Is that really the best we can hope for with libccd?
</blockquote></details>

BTW Not all of the tests fail at this precision -- I'd like a note/todo indicating why this value is so high and which combination of spheres is ending so badly.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 100 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD506u-7JF_tTN29Hb:-LDHoz_R0p5D86OX74ed:b-70drlg) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L100)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Without the bug fix on `_ccdDist`, one of these checks would fail. So I think it still makes sense to test both orders here?
</blockquote></details>

Refined f2f

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-:-LDHpG5C-Qecu2ERiShE:bbbhgqg)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: 4 of 7 files reviewed at latest revision, 14 unresolved discussions.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1157 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHOkq9DqlFcPJA0IuZ:-LDI0-Sw4R7ZD8FUSs01:bu6ttcl) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1157)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I have a bunch of concerns about this documentation

1. p1 and p2 are expressed in which frame? (Comment applies below as well).
2. The return value is undocumented
3. The relationship between the objects and the simplex isn't explained.
4. The semantics of the simplex isn't clear -- is it properly initialized upon invocation? Or is it an output value?

I know it's asking a lot, but if this were elevated to a Drake-quality piece of documentation, we'd all benefit.

Something along the lines of:

> Computes the distance between two non-penetrating convex objects, `obj1` and `obj2` based on the warm-started witness `simplex`, returning the distance and nearest points on each object.
@param obj1     The first convex object.
@param obj2     The second convex object.
@param ccd      The libccd configuration.
@param simplex   A witness to the objects' separation generated by the GJK algorithm. NOTE: the simplex is not necessarily sufficiently refined to report the *actual* distance and may be further refined in this method.
@param p1       If the objects are non-penetrating, the point on the surface of `obj1` closest to `obj2` (expressed in the world frame).
@param p2       If the objects are non-penetrating, the point on the surface of `obj2` closest to `obj1` (expressed in the world frame).
@returns The minimum distance between the two objects. If they are penetrating, -1 is returned.

(With appropriate `//` injected at appropriate line width, of course.)

Unknowns: What happens if the simplex is not initialized? Or it's lying?


</blockquote></details>

Done. Thanks a lot for writing this clear documentation!

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1160 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEK9Vw2zUMmP6lTio_:-LDI09KSA8ldkLxoNlyO:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1160)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI you could use backticks for variables rather than `@p` -- somewhat easier to read and would also work in Doxygen. (Irrelevant for this non-doxygen comment)
</blockquote></details>

Done.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1171 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDELcj-8lLlLxru7VT3:-LDI0qcVD-Hs7svHPtnr:b-sd4k07) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1171)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider moving these declarations down to just prior to their use and in the loop scope where possible. I had trouble figuring out what was used where by scanning the code. I believe `iterations` is just a local loop variable so can be moved to the for statement. Looks like only `closest_p` needs to be defined at the top?
</blockquote></details>

Done. I think `last_dist` needs to be defined at the top, since it is being updated in every iteration within the for loop.

---

*[include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h, line 1339 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHP_pY4uPAbZtKqK_B:-LDI4nrF0z2sd9rEjv-K:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/include/fcl/narrowphase/detail/convexity_based_algorithm/gjk_libccd-inl.h#L1339)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I'd modify this to reflect the comment on the other function. The difference between the two is that this does the warm starting of the libccd GJK implementation.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 105 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD0Cu-8dFS1VzoMN-w:-LDI9pn19fWb8auVhUG6:b-wrcts7) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/test_fcl_sphere_sphere.cpp#L105)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI If you expressed these in terms of your parameters it would be less brittle.

Then you could position the spheres in non-axis-aligned positions and still get the right answers.

</blockquote></details>

I think now with `SphereSpecification`, I can freely put the sphere anywhere?

---

*[test/test_fcl_sphere_sphere.cpp, line 35 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDENDqHDkC6alJkY4aC:-LDI4tqR8C5fC-LMDmsf:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L35)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI could add a space before closing comment.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 49 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHjNxQBqydRx57skZ8:-LDI8_OyEsSN3ENEMgri:bw08s3b) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L49)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I just realized: if you had spheres A and B and radius_A and radius_B, that makes p_FB and p_FA fit better.
</blockquote></details>

I am hesitant to do that, since the letter `A` will be overloaded, to mean both the sphere and the center of the sphere (as used in `p_FA`)

---

*[test/test_fcl_sphere_sphere.cpp, line 60 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDENTanAwLfe362WJEo:-LDI9Vs4EzfNza7s4ppi:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L60)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW consider a brief explanation of this notation and a link to Drake's documentation for it: http://drake.mit.edu/doxygen_cxx/group__multibody__spatial__pose.html

Likely FCL programmers are not familiar with it yet.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 116 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmRrY-eJALYDQKY-W:-LDI9b_V6mPuld1tnDvS:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L116)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW missing space before `-` operator.
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 123 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmaTb0L_IXDTLuG1A:-LDI9mkj2djBNs5ouFjn:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L123)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW These three comments are slightly redundant. The first one includes the next two. So, I'd suggest one of two options:

1. Delete the last two comments, or
2. delete the first comment and then move the last comment to between the two Vector declarations (i.e., comment, object, comment, object.)
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 143 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHnHi77nFbCXuFLsT-:-LDIAEYf9zKide9KCJCc:b35wj23) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L143)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This is a dangerous test -- this should be a `ASSERT_GT(min_distance, 0)`. Otherwise, the spheres could be changed to be in collision and *no* testing would take place. Obviously, this will require you to guarantee that none of the sphere pairs are colliding.  :)
</blockquote></details>

Done. Good point!

---

*[test/test_fcl_sphere_sphere.cpp, line 180 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDEOkIVE9veSRoCuxRv:-LDIAc_nDhYlU6QRiEq3:bl00uvk) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L180)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Not all of the tests fail at this precision -- I'd like a note/todo indicating why this value is so high and which combination of spheres is ending so badly.
</blockquote></details>

Done. All pairs of combination are bad.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 100 at r1](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDD506u-7JF_tTN29Hb:-LDICUCF0yKFLWQLM6tV:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/e12e04828fc2b05cbe4457b2eaf1c12eb404f4e5/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L100)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Refined f2f
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[test/test_fcl_sphere_sphere.cpp, line 156 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDIfOBXB-FrJMRZUBMp:-LDIfOBXB-FrJMRZUBMq:b42z4mt) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L156)):*
> ```cpp
>             sphere_sphere_distance.min_distance, sphere_sphere_distance.p_S2P2,
>             sphere_sphere_distance.p_S1P1, tol);
>         
> ```

BTW some odd characters at beginning of line here?

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> It's worth noting that this PR seems to have fixed the persistent mac CI failure we've had. That's outstanding! Now I just have to fix the box-box test so that it passes in the mac release.

---

Reviewed 2 of 3 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 47 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDMTxTK7j0-T3hHSpba:-LDMTxTK7j0-T3hHSpbb:b-h8rj6j) ([raw file](https://github.com/flexible-collision-library/fcl/blob/9eb68d3ae3a788516406ca81e631bc0be9f1068b/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L47)):*
> ```cpp
> 
> // Given two spheres, sphere 1 has radius1, and centered at point A, whose
> // position is p_FA measured and expressed in frame A; sphere 2 has radius2,
> ```

BTW measured and expressed in frame __*F*__? Certainly, according to Drake's monogram notation `p_FA` is an abbreviation for `p_FA_F`.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 126 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDMUcqv4WSth5vNBK7j:-LDMUcqv4WSth5vNBK7k:b262p0k) ([raw file](https://github.com/flexible-collision-library/fcl/blob/9eb68d3ae3a788516406ca81e631bc0be9f1068b/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L126)):*
> ```cpp
>   for (int i = 0; i < static_cast<int>(spheres.size()); ++i) {
>     for (int j = i + 1; j < static_cast<int>(spheres.size()); ++j) {
>       if ((spheres[i].center - spheres[j].center).norm() >
> ```

BTW This is a repeat of the other test's issue -- if someone changes the spheres such that they collide, no tests will be executed for that pair and the test will meaninglessly pass. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-:-LDMV3eu1ItSW52F8W2Y:b-b6ukxe)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 1 of 3 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[test/test_fcl_sphere_sphere.cpp, line 117 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDHmWN67BGJK5AFIJgm:-LDN7y7iC04OFdfKP3Bd:b-896fix) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L117)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI *slightly* more efficient to negate the radius. But yours reads better. :)
</blockquote></details>

Done.

---

*[test/test_fcl_sphere_sphere.cpp, line 156 at r2](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDIfOBXB-FrJMRZUBMp:-LDN7yeL1JTd0K39XUNr:b-n0jcoy) ([raw file](https://github.com/flexible-collision-library/fcl/blob/c6a4ba73942e9fef8d289d37d9f6d22237d5d4bf/test/test_fcl_sphere_sphere.cpp#L156)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW some odd characters at beginning of line here?
</blockquote></details>

Hmm I do not see them, maybe a reviewable glitch?

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 47 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDMTxTK7j0-T3hHSpba:-LDN5Vmf3MJFJAZ51NCQ:b-376vyb) ([raw file](https://github.com/flexible-collision-library/fcl/blob/9eb68d3ae3a788516406ca81e631bc0be9f1068b/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L47)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW measured and expressed in frame __*F*__? Certainly, according to Drake's monogram notation `p_FA` is an abbreviation for `p_FA_F`.
</blockquote></details>

Done. Good catch.

---

*[test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp, line 126 at r3](https://reviewable.io/reviews/flexible-collision-library/fcl/290#-LDMUcqv4WSth5vNBK7j:-LDN7m0N8EstP6u3vjJZ:b-yh10tr) ([raw file](https://github.com/flexible-collision-library/fcl/blob/9eb68d3ae3a788516406ca81e631bc0be9f1068b/test/narrowphase/detail/convexity_based_algorithm/test_gjk_libccd-inl.cpp#L126)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This is a repeat of the other test's issue -- if someone changes the spheres such that they collide, no tests will be executed for that pair and the test will meaninglessly pass. 
</blockquote></details>

Done. Good catch

---


*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/flexible-collision-library/fcl/290)*
<!-- Sent from Reviewable.io -->
